### ISSUE TITLE
[FEATURE] More control over explicit test example sources

### ISSUE THREAD

**Issue by IvanRibakov** [schemathesis/schemathesis#2369]
### Is your feature request related to a problem? Please describe

Right now if OpenAPI endpoint has an example under `requestBody` AND request schema has an example value for each field, Schemathesis generates 2 separate requests with different payloads but the same request parameters. This is problematic for endpoints that create resources and accept resource IDs via request parameters, for example:
```
POST /book/{book_id}
```

because the 2nd request typically fails to create the resource due to the ID clash (that is part of the request path and doesn't change from example to example).

### Describe the solution you'd like

Ideally, something like this might take care of a wide range of needs:

- New CLI flag `--explicit-example-count` with values:
    - `single` - at most one explicit request per operation may be performed
    - `multiple` (default) - more than one request per operation may be performed
- New CLI flag `--explicit-example-source` with values:
  - `all`
  - `operation` - only examples defined under operation's `requestBody > content > {type} > example/examples` are used
  - `schema` - only example(s) constructed from the operation's request schema's individual field examples (not sure if more than one example can be constructed from the schema definition)

This approach can be later extended to allow filtering by example name when using [`examples` with named values](https://swagger.io/specification/#media-type-object).

### Describe alternatives you've considered

1. Removing either operation example or field example
    - Not ideal option as both example types typically play role when rendering the API docs from OpenAPI schema. If possible we'd like to keep both field-level examples and operation-level examples
2. Use hooks
    - Could work but would significantly increase the complexity of the API testing. Currently we rely on the "create" operation test to set up test data that is later used by "read", "update" and "delete" operation tests. If we started using hooks to delete each created record after the test case completed, we'd then also need to start setting up the test data for "read", "update" and "delete"operation tests.

**Comment by Stranger6667** [comment]
Hi!

I like the idea and the proposed solution, however, I don't have the capacity to work on it right now (probably in a few weeks). Additionally, the last part about create / read / update / delete somehow reminds me of stateful testing - have you considered using that approach as well? However, the approach you are describing goes more toward explicit examples rather than generated values.

**Comment by IvanRibakov** [comment]
> Additionally, the last part about create / read / update / delete somehow reminds me of stateful testing - have you considered using that approach as well?

Fair point. We've tried using stateful tests in the past on an earlier version of schemathesis but we ran into some other issues. I don't recall the details now as more than a year has passed.

> However, the approach you are describing goes more toward explicit examples rather than generated values.

Are you saying that stateful tests are not supported by the explicit phase? I did see some mention of the data generation in the stateful docs but not 100% if that means stateful tests can not be used in other phases. If so, that would indeed be a limiting factor for us to proceed in this direction.

**Comment by Stranger6667** [comment]
I am reluctant to extend the API surface for this use case as it looks solvable via the `before_add_examples` hook, which receives a mutable list of `Case` objects before they're executed as explicit examples. You can trim it to a single example per operation:

```python
import schemathesis

@schemathesis.hook
def before_add_examples(ctx, examples):
    if examples:
        del examples[1:]
```

Or if you only want operation-level examples (ignoring the ones constructed from field-level example values in the schema), you can inspect the case body/parameters to decide which to keep, or simply limit to the first one as above.

This hooks globally; to scope it to a specific schema or test function, see the [hooks documentation](https://schemathesis.readthedocs.io/en/stable/extending.html#hooks).

Given this covers your use case (I think), I'll close this. If you hit something the hook can't express, feel free to reopen.

Otherwise, data generation now uses explicit examples too, including the stateful phase + "examples" phases parses responses and uses resource ids in subsequent requests.
